staging/vhost-device-gpu/src/device.rs

```diff
@@ -0,0 +1,1628 @@
+                    x,
+                    y,
+                    z,
+                    w,
+                    h,
+                    d,
+                    level,
+                    stride,
+                    layer_stride,
+                    offset,
+                };
+
+                virtio_gpu.transfer_write(ctx_id, resource_id, transfer)
+            }
+            GpuCommand::TransferFromHost3d(virtio_gpu_transfer_host_3d {
+                box_: virtio_gpu_box { x, y, z, w, h, d },
+                offset,
+                resource_id,
+                level,
+                stride,
+                layer_stride,
+            }) => {
+                let ctx_id = hdr.ctx_id;
+
+                let transfer = Transfer3D {
+                    x,
+                    y,
+                    z,
+                    w,
+                    h,
+                    d,
+                    level,
+                    stride,
+                    layer_stride,
+                    offset,
+                };
+
+                virtio_gpu.transfer_read(ctx_id, resource_id, transfer, None)
+            }
+            GpuCommand::CmdSubmit3d {
+                fence_ids,
+                mut cmd_data,
+            } => virtio_gpu.submit_command(hdr.ctx_id, &mut cmd_data, &fence_ids),
+            GpuCommand::ResourceCreateBlob(_info) => {
+                panic!("virtio_gpu: GpuCommand::ResourceCreateBlob unimplemented");
+            }
+            GpuCommand::SetScanoutBlob(_info) => {
+                panic!("virtio_gpu: GpuCommand::SetScanoutBlob unimplemented");
+            }
+            GpuCommand::ResourceMapBlob(_info) => {
+                panic!("virtio_gpu: GpuCommand::ResourceMapBlob unimplemented");
+            }
+            GpuCommand::ResourceUnmapBlob(_info) => {
+                panic!("virtio_gpu: GpuCommand::ResourceUnmapBlob unimplemented");
+            }
+        }
+    }
+
+    fn process_queue_chain(
+        &mut self,
+        virtio_gpu: &mut impl VirtioGpu,
+        vring: &VringRwLock,
+        head_index: u16,
+        reader: &mut Reader,
+        writer: &mut Writer,
+        signal_used_queue: &mut bool,
+    ) -> Result<()> {
+        let mut response = ErrUnspec;
+        let mem = self.mem.as_ref().unwrap().memory().into_inner();
+
+        let ctrl_hdr = match GpuCommand::decode(reader) {
+            Ok((ctrl_hdr, gpu_cmd)) => {
+                // TODO: consider having a method that return &'static str for logging purpose
+                let cmd_name = format!("{:?}", gpu_cmd);
+                let response_result = self.process_gpu_command(virtio_gpu, &mem, ctrl_hdr, gpu_cmd);
+                // Unwrap the response from inside Result and log information
+                response = match response_result {
+                    Ok(response) => response,
+                    Err(response) => {
+                        debug!("GpuCommand {cmd_name} failed: {response:?}");
```

Comment:
```suggestion
                        debug!("GpuCommand {cmd_name:?} failed: {response:?}");
```
and you could drop the cmd_name?